protocols/autonat/src/v2/server/handler/dial_request.rs

```diff
@@ -0,0 +1,349 @@
+            return Poll::Ready(ConnectionHandlerEvent::NotifyBehaviour(Either::Right(
+                status_update,
+            )));
+        }
+        Poll::Pending
+    }
+
+    fn on_behaviour_event(&mut self, _event: Self::FromBehaviour) {}
+
+    fn on_connection_event(
+        &mut self,
+        event: ConnectionEvent<
+            Self::InboundProtocol,
+            Self::OutboundProtocol,
+            Self::InboundOpenInfo,
+            Self::OutboundOpenInfo,
+        >,
+    ) {
+        match event {
+            ConnectionEvent::FullyNegotiatedInbound(FullyNegotiatedInbound {
+                protocol, ..
+            }) => {
+                if self
+                    .inbound
+                    .try_push(start_handle_request(
+                        protocol,
+                        self.observed_multiaddr.clone(),
+                        self.client_id,
+                        self.dial_back_cmd_sender.clone(),
+                        self.status_update_sender.clone(),
+                        self.rng.clone(),
+                    ))
+                    .is_err()
+                {
+                    tracing::warn!(
+                        "failed to push inbound request handler, too many requests in flight"
+                    );
+                }
+            }
+            ConnectionEvent::ListenUpgradeError(ListenUpgradeError { error, .. }) => {
+                tracing::debug!("inbound request failed: {:?}", error);
+            }
+            _ => {}
+        }
+    }
+}
+
+enum HandleFail {
+    InternalError(usize),
+    RequestRejected,
+    DialRefused,
+    DialBack {
+        idx: usize,
+        result: Result<(), DialBackStatus>,
+    },
+}
+
+impl From<HandleFail> for DialResponse {
+    fn from(value: HandleFail) -> Self {
+        match value {
+            HandleFail::InternalError(addr_idx) => Self {
+                status: ResponseStatus::E_INTERNAL_ERROR,
+                addr_idx,
+                dial_status: DialStatus::UNUSED,
+            },
+            HandleFail::RequestRejected => Self {
+                status: ResponseStatus::E_REQUEST_REJECTED,
+                addr_idx: 0,
+                dial_status: DialStatus::UNUSED,
+            },
+            HandleFail::DialRefused => Self {
+                status: ResponseStatus::E_DIAL_REFUSED,
+                addr_idx: 0,
+                dial_status: DialStatus::UNUSED,
+            },
+            HandleFail::DialBack { idx, result } => Self {
+                status: ResponseStatus::OK,
+                addr_idx: idx,
+                dial_status: match result {
+                    Err(DialBackStatus::DialErr) => DialStatus::E_DIAL_ERROR,
+                    Err(DialBackStatus::DialBackErr) => DialStatus::E_DIAL_BACK_ERROR,
+                    Ok(()) => DialStatus::OK,
+                },
+            },
+        }
+    }
+}
+
+async fn handle_request_internal<I>(
```

Comment:
Please order these functions in call order, i.e. top-most function calls 2nd one etc.

Reply:
did that